Add --on-error to all launched mons, pass name of process.

flexd opened the pull request:
Related to pull request on mon https://github.com/visionmedia/mon/pull/17

This lets us have a -on-error script for all mon processes that mongroup starts.

Mongroup passes the name of the process as the first argument here, and with my modifications to mon , mon passes the process PID.

That way you can use $1 (the first argument) as the process name and $2 (the second argument) as the PID in your --on-error script.

I also added an example that uses /usr/bin/mail to send a email using the variables, I use this to send alerts to NotifyMyAndroid (NMAMail) so I get alerts on my phone like the one in the mon pull-request :)

At the moment this assumes/forces you to have a -on-error script.  It might be smart to only add the -on-error flag to the command in start_proc if $ERRORCMD is not null, and then not set a default.

That way you can use mongroup without an --on-error event, I have not implemented that yet. :)

flexd committed:
This fixes a lot of issues talked about in #5

Change config-format to solve some problems.
This commit changes how the config format works.
We take advantage of bash's built-in file sourcing
abilities to allow us to have a config file without
reading a file like before. This solution is more
robust, and it fixes several flaws with the old system
mainly:
  * You can have commands with = in them, something that would break the
  * old system. And you we also no longer break if there is a whitespace
  * in our command :)

flexd committed:
Pass the name of the process to --on-error

flexd committed:
Add example and add default.

flexd commented:
I'm working on making the --on-error and --on-restart flags optional, as well as cleaning up the code a bit :-)

flexd committed:
Change config-format to solve some problems.
This commit changes how the config format works.
We take advantage of bash's built-in file sourcing
abilities to allow us to have a config file without
reading a file like before. This solution is more
robust, and it fixes several flaws with the old system
mainly:
  * You can have commands with = in them, something that would break the
  * old system. And you we also no longer break if there is a whitespace
  * in our command :)

flexd committed:
Allows us to only apply flags when needed.

flexd committed:
Updated update() too.

flexd commented:
These should probably be rebased and be given some proper commit messages before being merged, if you choose to do so. I do hope so though :-)

I changed the format of the config file for several reasons pointed out to me by the fine people over in #bash on freenode.

You would break the script if you had a = in the command, and you would break the script if you had a whitespace in the command, and I'm sure there are more.

You can specify the --on-error and --on-restarts commands to be run in the config, (which is now just a normal bash script that is being sourced by the way)
by doing this:

on_error=(./on_error)
on_restart=(./on_restart)

You then specify processes pretty much like normal, they are passed as bash arrays since using associative arrays would look messy.

web=(node server.js 8001)

and finally (unfortunately) you have to pass an array called processes of all the processes to run.

processes=( web )

You can also pass --attempts, like so

attempts=3

I also took the liberty of changing all the variables to lower-case, you should only use uppercase if they are environmental variables.

There is still lots of stuff that could be cleaned up (mainly around start_proc for now) but it's 3:23 AM here so I am going to bed :)

tj commented on bin/mongroup:
```diff
@@ -53,35 +52,27 @@ abort() {
   exit 1
 }
+# working: mon -p ./pids/web.pid -m ./pids/web.mon.pid -l ./logs/web.log --on-error "./on_error.sh web" "node server.js 8001"
+# not working: mon -p ./pids/web.pid -m ./pids/web.mon.pid -l ./logs/web.log --on-error ./on_error.sh web node server.js 8001
```
this is pretty ambiguous anyway, quoting makes more sense

flexd replied:
Yeah, that was more me testing things. I was having trouble with it either trying to execute 'web' or just the first part of the command 'node'

tj commented on bin/mongroup:
```diff
@@ -100,10 +91,24 @@ set_config() {
+    -l "$log_dir/$1.log")
+
+  if [ -n "$attempts" -a "$attempts" != "" -a -f "$attempts" ]
+  then
```
these seem a big unwieldy, it would be great if we had something less specific for passing options through without handling each individual flag. Also `; then` style like above is nicer IMO less clutter but that's just me being OCD

flexd replied:
We could at least make the checking bit a function, that way it would be a bit more DRY?

I haven't really used much bash scripting before, I'll try to make it a bit better once I'm done with my exams this week.

tj replied:
sounds good ill try and pull this down and have a look soonish as well

flexd committed:
Load config once (twice at most) instead of often.

flexd committed:
Load config once (twice at most) instead of often.

Less messy, double brackets everything, comments!

Added a useful on_error example.

Renamed the on_error.sh, it is not a sh script, so .sh is silly

Add a clean method to remove logfiles.

flexd committed:
Less messy, double brackets everything, comments!

flexd committed:
Added a useful on_error example.

flexd commented:
Added some comments as I understood more of the code.

Double brackets is the "new test" in bash, some of the smaller improvements are: You can use && and || instead of -a and -o, and you do not need to "$variable" expand variables when testing like that. Looks a bit neater.

I also changed the script to only read the config once, or twice if you use --config.

I'm sure there are more things that could be cleaned up. I'll try to find some more things as well as rebase this all into a nicer commit after my exam on thursday.

jgallen23 commented:
awesome, nice work.  Once you clean up the commit, I'll merge it.

flexd committed:
All commands accept multiple targets, fixes #6

Start, stop, restart and log all work.

Let me know if you spot any bugs or problems.

flexd committed:
Merge branch 'patch-1'

flexd committed:
Merge branch 'master' of https://github.com/flexd/mongroup

Conflicts:
	bin/mongroup
	example/mongroup.conf

flexd committed:
Pesky rebases

flexd commented:
I sometimes hate git.  Been struggeling with getting the damned rebasing to work properly. Still a lot of commits.

Do you want them all as one?

By the way, I made start, stop, restart and log all accept multiple commands.

![Starting and stopping](http://cognitive.io/mongroup/start_stop.png)

![Already started or stopped](http://cognitive.io/mongroup/already_running_or_stopped.png)

The only thing missing at the moment is mongroup telling you that you tried to start/stop/restart/log a process that does not exist. It will just return nothing at the moment.

flexd commented:
Err, restart is broken (it stops it, then tells us its already running, doh).

I somehow introduced a bug stopping processes, I'll sort it out later :-)

tj commented:
dont worry about the # of commits for now, we can squash it all into one so it's cleaner

flexd commented:
Meh, going to bed. Not sure what is going on with stop(). The mon process is stopping, but its leaving the monitored process hanging/still running.

Here's a log if anyone feels like debugging. 

https://gist.github.com/0b193e666a307c2de7a0

It kind of looks like mon is the one misbehaving there.

Pushed my latest changes in case anyone feels like looking while I sleep.

flexd committed:
WIP

flexd committed:
Fixed the stop() bug.

flexd committed:
Fixed the handle_restart() bug.

flexd commented:
![Restarting](http://cognitive.io/mongroup/restart.png)

flexd committed:
Fix typo in README and remove on_error with .sh extension

flexd commented:
I am fairly certain everything works as intended now.

You can start, stop and restart all, a single or multiple processes.

I was thinking of perhaps allowing regex/wildcards as well, so you could do mongroup restart web\* or something like that?

flexd commented:
Since I have some time on my hands today, what do we actually need to get this merged? I'm not great at squashing commits into one. I think perhaps some small changes to make the message style uniform could be nice.
